# 📦 Prepare `dataset/` at repo root (robust, repo-root aware)

This notebook downloads **dataset.zip** from Google Drive, extracts it to  
`<repo>/dataset/` (`soccer_vision/dataset`), fixes any nested paths created by the
notebook’s working directory, and validates the expected structure:

```

soccer_vision/
├── dataset/
    ├──train/
        ├── images/
        ├── labels/
    ├── val/           (optional; can be created later by the training script)
```

## Install/imports

In [1]:
import os, zipfile, shutil, sys
from pathlib import Path

# Install gdown if needed
try:
    import gdown  # noqa: F401
except Exception:
    %pip -q install gdown
import gdown

## Configuration

- `ROOT_NAME`: the folder name of your repo.
- `GDRIVE_DATASET_ID`: the Google Drive **file id** for `dataset.zip`.

In [2]:
ROOT_NAME = "soccer_vision"  # repo folder name
GDRIVE_DATASET_ID = "13c8_NnjszJBDuLAHCr8jKq3eBP7Pocez"  # <-- your dataset.zip file id

## Helper functions

- `resolve_repo_root()` finds the repo root no matter where this notebook runs.
- `dataset_looks_ok()` checks expected folders.
- `tree()` prints a small directory tree for quick inspection.

In [3]:
def resolve_repo_root() -> Path:
    """Find the repo root named ROOT_NAME, regardless of where this notebook runs."""
    env = os.getenv("SOCCER_VISION_ROOT")
    if env:
        p = Path(env).expanduser().resolve()
        if p.name != ROOT_NAME:
            print(f"[WARN] SOCCER_VISION_ROOT is set to {p}, but folder name != {ROOT_NAME}")
        return p

    here = Path.cwd().resolve()
    if here.name == ROOT_NAME and (here / "notebooks").exists():
        return here
    if here.name == "notebooks" and here.parent.name == ROOT_NAME:
        return here.parent
    for p in [here, *here.parents]:
        if p.name == ROOT_NAME:
            return p
    for p in [here, *here.parents]:
        if (p / "notebooks").exists():
            return p
    return here

def dataset_looks_ok(ds: Path) -> bool:
    return (ds / "train" / "images").exists() and (ds / "train" / "labels").exists()

def tree(path: Path, max_files=5):
    print(f"\n[Tree] {path.resolve()}")
    for root, dirs, files in os.walk(path):
        indent = "  " * (len(Path(root).relative_to(path).parts))
        print(f"{indent}{Path(root).name}/")
        for d in sorted(dirs):
            print(f"{indent}  {d}/")
        for i, f in enumerate(sorted(files)):
            if i >= max_files:
                print(f"{indent}  ... (+{len(files)-max_files} more)")
                break
            print(f"{indent}  {f}")


## Locate repo root & define key paths

In [4]:
from soccer_vision.notebooks.modules.paths import find_repo_root

BASE = find_repo_root()
print(f"[BASE] {BASE}")

DATASET_DIR = BASE / "dataset"       # expected final location
DATASET_ZIP = BASE / "dataset.zip"   # temporary zip path

[BASE] Z:\Proyectos\Robotica\footbot\soccer_vision


## Fix a previously nested location

If a prior run placed the dataset under `notebooks/soccer_vision/dataset`, this
moves it back to `<repo>/dataset`.

In [5]:
nested = BASE / "notebooks" / ROOT_NAME / "dataset"
if nested.exists() and not DATASET_DIR.exists():
    print(f"[FIX] Moving nested dataset -> {DATASET_DIR}")
    DATASET_DIR.parent.mkdir(parents=True, exist_ok=True)
    shutil.move(str(nested), str(DATASET_DIR))
    # Attempt to remove the now-empty wrapper dir
    try:
        shutil.rmtree((BASE / "notebooks" / ROOT_NAME), ignore_errors=True)
    except Exception as e:
        print("[WARN] Could not remove nested wrapper folder:", e)

## Download `dataset.zip` from Google Drive (only if missing)

Requires the file to be shared as **Anyone with the link**.

In [6]:
if DATASET_DIR.exists() and dataset_looks_ok(DATASET_DIR):
    print("[INFO] dataset/ already exists, skipping download.")
else:
    print("[DL] Downloading dataset.zip …")
    url = f"https://drive.google.com/uc?id={GDRIVE_DATASET_ID}"
    ok = gdown.download(url, str(DATASET_ZIP), quiet=False)
    if not ok:
        sys.exit("[ERR] gdown failed. Make sure the Drive file is public (Anyone with the link).")

[INFO] dataset/ already exists, skipping download.


## Unzip & normalize to `<repo>/dataset`

In [7]:
if not (DATASET_DIR.exists() and dataset_looks_ok(DATASET_DIR)):
    print("[UNZIP] Extracting …")
    with zipfile.ZipFile(DATASET_ZIP, "r") as z:
        z.extractall(BASE)

    if not dataset_looks_ok(DATASET_DIR):
        # Find a folder that contains train/images and train/labels and move it to dataset/
        candidates = []
        for p in BASE.iterdir():
            if p.is_dir() and (p / "train" / "images").exists() and (p / "train" / "labels").exists():
                candidates.append(p)
        if not candidates:
            sys.exit("[ERR] Could not locate a folder with train/images and train/labels after unzip.")

        pick = next((c for c in candidates if c.name == "dataset"), candidates[0])
        if pick != DATASET_DIR:
            if DATASET_DIR.exists():
                shutil.rmtree(DATASET_DIR)
            shutil.move(str(pick), str(DATASET_DIR))
    
    print("[UNZIP] Finished")

## Clean up temporary files

In [8]:
print("[CLEAN] Deleting dataset.zip (if present)")
try:
    DATASET_ZIP.unlink()
except FileNotFoundError:
    pass

[CLEAN] Deleting dataset.zip (if present)


## Final validation

We should now have `soccer_vision/dataset/train/images` and `soccer_vision/dataset/train/labels`.

In [9]:
if not dataset_looks_ok(DATASET_DIR):
    sys.exit("[ERR] dataset/ is missing expected subfolders (train/images and train/labels).")

print("[OK] Dataset ready at:", DATASET_DIR)
tree(DATASET_DIR)

[OK] Dataset ready at: Z:\Proyectos\Robotica\footbot\soccer_vision\dataset

[Tree] Z:\Proyectos\Robotica\footbot\soccer_vision\dataset
dataset/
  train/
  val/
  classes.txt
  notes.json
  train/
    images/
    labels/
    labels.cache
    images/
      00b200b7-photo_2025-10-12_16-30-16_aug_2.jpg
      01aef225-photo_2025-10-12_16-30-49_aug_2.jpg
      02147359-photo_2025-10-12_22-01-54_aug_3.jpg
      02260e77-photo_2025-10-12_22-01-58_aug_1.jpg
      0392df9f-photo_2025-10-12_16-28-24_aug_3.jpg
      ... (+455 more)
    labels/
      00b200b7-photo_2025-10-12_16-30-16_aug_2.txt
      01aef225-photo_2025-10-12_16-30-49_aug_2.txt
      02147359-photo_2025-10-12_22-01-54_aug_3.txt
      02260e77-photo_2025-10-12_22-01-58_aug_1.txt
      0392df9f-photo_2025-10-12_16-28-24_aug_3.txt
      ... (+455 more)
  val/
    images/
    labels/
    labels.cache
    images/
      0cd54e92-photo_2025-10-12_22-02-19_aug_1.jpg
      16444e52-photo_2025-10-12_22-01-24_aug_2.jpg
      1b18bdc6-photo_20

## Install Ultralytics (YOLO)
Only needed the first time you run this environment.

In [10]:
%pip -q install ultralytics torch

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 23.2.1 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


## Training parameters
Adjust these, then run the next cell to retrain.


In [11]:
MODEL_BACKBONE   = "yolo11s.pt"
EPOCHS           = 60
IMG_SIZE         = 640
BATCH_SIZE       = 16
DEVICE           = "0"            # GPU id like "0" or "cpu"
TRAIN_PCT        = 0.90           # used only if val/ is empty and --no-split is False
NO_SPLIT         = False          # set True if you don't want main.py to create/move a val split
PATIENCE         = 20             # early stopping patience
WORKERS          = 8              # dataloader workers
SEED             = 0              # for reproducible split & training seed
RUN_NAME         = None           # if None -> defaults to "<backbone_stem>_train"
OUT_SUBDIR       = None           # if None -> defaults to "<backbone_stem>" under soccer_vision/models
FINAL_WEIGHTS    = "soccer_yolo.pt"  # file name saved in models/<OUT_SUBDIR>/
CONF_FOR_VALPREV = 0.25           # quick val prediction confidence


## Retrain the model
This calls the project script so all runs/models land **inside** `soccer_vision/`.

In [12]:
# Make sure we can import notebooks.modules.* no matter the current CWD
from pathlib import Path
import sys

here = Path.cwd().resolve()
repo = here if (here.name == "soccer_vision") else (here.parent if here.name == "notebooks" else None)
if not repo:
    # walk up until we find soccer_vision/
    for p in [here, *here.parents]:
        if (p / "soccer_vision" / "notebooks" / "modules").exists():
            repo = p / "soccer_vision"
            break

assert repo and (repo / "notebooks" / "modules").exists(), f"Could not find modules under {here}"
if str(repo) not in sys.path:
    sys.path.insert(0, str(repo))  # add repo root to sys.path

print("[REPO]", repo)

[REPO] Z:\Proyectos\Robotica\footbot\soccer_vision


In [13]:
# Live logs training
from soccer_vision.notebooks.modules.logging_utils import get_logger, log
from soccer_vision.notebooks.modules.paths import find_repo_root
from soccer_vision.notebooks.modules.train import train_yolo

# Optional: INFO (or DEBUG) level
get_logger()  # already INFO by default

BASE = find_repo_root()
SV_DIR = BASE  # shorthand

# ---- knobs to tweak from notebook ----
MODEL_BACKBONE   = SV_DIR / "yolo11s.pt"   # or "yolo11s.pt" to auto-download
EPOCHS           = 60
IMG_SIZE         = 640
BATCH_SIZE       = 16
DEVICE           = "0"                     # "0" for first GPU; "cpu" to force CPU
WORKERS          = 8
SEED             = 0
PATIENCE         = 20
TRAIN_PCT        = 0.90
COPY_SPLIT       = False
NO_SPLIT         = False
FINAL_WEIGHTS    = "soccer_yolo.pt"
CONF_FOR_VALPREV = 0.25
RUN_NAME         = "yolo11s_train"
OUT_SUBDIR       = "yolo11s"               # models/<OUT_SUBDIR>

result = train_yolo(
    base_dir=SV_DIR,
    model=MODEL_BACKBONE,
    epochs=EPOCHS,
    imgsz=IMG_SIZE,
    batch=BATCH_SIZE,
    device=DEVICE,
    workers=WORKERS,
    seed=SEED,
    patience=PATIENCE,
    train_pct=TRAIN_PCT,
    copy_split=COPY_SPLIT,
    final_name=FINAL_WEIGHTS,
    conf_for_valprev=CONF_FOR_VALPREV,
    run_name=RUN_NAME,
    out_subdir=OUT_SUBDIR,
    no_split=NO_SPLIT,
    logger=log,
)

print("\n--- RESULT ---")
print("Best Weights :", result.best_weights)
print("Run Dir      :", result.run_dir)
print("Artifacts    :", result.artifacts_dir)
print("Val Predicts :", result.val_pred_dir)
print("Device Used  :", result.device_used)


[INFO] [INFO] Validation split already present. Skipping split.
[INFO] [OK] Wrote Z:\Proyectos\Robotica\footbot\soccer_vision\data.yaml
[INFO] [INFO] ULTRALYTICS_HOME=Z:\Proyectos\Robotica\footbot\soccer_vision
[INFO] [TRAIN] model=Z:\Proyectos\Robotica\footbot\soccer_vision\yolo11s.pt epochs=60 imgsz=640 batch=16 device=0 workers=8 seed=0 patience=20
New https://pypi.org/project/ultralytics/8.3.221 available  Update with 'pip install -U ultralytics'
Ultralytics 8.3.0  Python-3.11.5 torch-2.5.1+cu121 CUDA:0 (NVIDIA GeForce RTX 3050 Ti Laptop GPU, 4096MiB)
engine\trainer: task=detect, mode=train, model=Z:\Proyectos\Robotica\footbot\soccer_vision\yolo11s.pt, data=Z:\Proyectos\Robotica\footbot\soccer_vision\data.yaml, epochs=60, time=None, patience=20, batch=16, imgsz=640, save=True, save_period=-1, cache=False, device=0, workers=8, project=Z:\Proyectos\Robotica\footbot\soccer_vision\runs, name=yolo11s_train, exist_ok=True, pretrained=True, optimizer=auto, verbose=True, seed=0, determinis

train: Scanning Z:\Proyectos\Robotica\footbot\soccer_vision\dataset\train\labels.cache... 460 images, 0 backgrounds, 0 corrupt: 100%|██████████| 460/460 [00:00
val: Scanning Z:\Proyectos\Robotica\footbot\soccer_vision\dataset\val\labels.cache... 52 images, 0 backgrounds, 0 corrupt: 100%|██████████| 52/52 [00:00<?, ?it


Plotting labels to Z:\Proyectos\Robotica\footbot\soccer_vision\runs\yolo11s_train\labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.001667, momentum=0.9) with parameter groups 81 weight(decay=0.0), 88 weight(decay=0.0005), 87 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 8 dataloader workers
Logging results to Z:\Proyectos\Robotica\footbot\soccer_vision\runs\yolo11s_train
Starting training for 60 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       1/60      4.24G      1.016      3.097       1.25         21        640: 100%|██████████| 29/29 [00:24<00:00,  1.20it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.49s/it]

                   all         52         69      0.727      0.685      0.738      0.484



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       2/60       4.3G     0.9665      1.195      1.192         19        640: 100%|██████████| 29/29 [00:28<00:00,  1.01it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.36s/it]

                   all         52         69      0.365      0.457      0.402      0.223



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       3/60      4.31G      1.022      1.038      1.233         32        640: 100%|██████████| 29/29 [00:44<00:00,  1.53s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:03<00:00,  1.61s/it]

                   all         52         69       0.25     0.0795     0.0898     0.0349



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       4/60       4.3G     0.9892     0.9698      1.198         24        640: 100%|██████████| 29/29 [00:46<00:00,  1.59s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:03<00:00,  1.81s/it]

                   all         52         69      0.613      0.396      0.452      0.254



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       5/60      4.28G     0.9629     0.8514      1.174         33        640: 100%|██████████| 29/29 [00:24<00:00,  1.16it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.19s/it]

                   all         52         69       0.67      0.441       0.51      0.346



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       6/60      4.31G     0.8799      0.752      1.115         27        640: 100%|██████████| 29/29 [00:28<00:00,  1.01it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.44s/it]

                   all         52         69      0.822      0.903      0.908      0.696



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       7/60       4.3G     0.8894     0.7338      1.133         25        640: 100%|██████████| 29/29 [00:29<00:00,  1.02s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.28s/it]

                   all         52         69      0.656      0.821      0.808      0.612



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       8/60       4.3G      0.846     0.7039       1.11         32        640: 100%|██████████| 29/29 [00:44<00:00,  1.53s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.50s/it]

                   all         52         69       0.75      0.825      0.926      0.701



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       9/60      4.28G     0.8609      0.691      1.096         31        640: 100%|██████████| 29/29 [00:42<00:00,  1.48s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.49s/it]

                   all         52         69      0.908      0.906      0.935      0.731



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      10/60      4.31G     0.8577     0.6731      1.115         23        640: 100%|██████████| 29/29 [00:38<00:00,  1.32s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.39s/it]

                   all         52         69      0.988      0.915      0.931      0.718



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      11/60      4.31G     0.7948     0.6164      1.071         32        640: 100%|██████████| 29/29 [00:51<00:00,  1.78s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:03<00:00,  1.51s/it]

                   all         52         69      0.966      0.933      0.969      0.808



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      12/60       4.3G     0.7704     0.6233      1.079         27        640: 100%|██████████| 29/29 [00:48<00:00,  1.68s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:03<00:00,  1.50s/it]

                   all         52         69      0.924      0.906      0.956       0.72



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      13/60       4.3G     0.7638     0.6031      1.058         28        640: 100%|██████████| 29/29 [00:43<00:00,  1.50s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:04<00:00,  2.04s/it]

                   all         52         69      0.979      0.988      0.974      0.798



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      14/60      4.32G     0.7371     0.5607      1.042         27        640: 100%|██████████| 29/29 [00:39<00:00,  1.36s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:03<00:00,  1.58s/it]

                   all         52         69      0.963      0.963      0.983      0.809



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      15/60       4.3G     0.7197     0.5328      1.041         33        640: 100%|██████████| 29/29 [00:45<00:00,  1.58s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:03<00:00,  1.62s/it]

                   all         52         69      0.995      0.977      0.978      0.811



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      16/60      4.31G     0.7176     0.5445      1.034         22        640: 100%|██████████| 29/29 [00:50<00:00,  1.74s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.48s/it]

                   all         52         69       0.93      0.896      0.943      0.782



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      17/60       4.3G     0.7062     0.5391      1.038         36        640: 100%|██████████| 29/29 [00:29<00:00,  1.00s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:03<00:00,  1.65s/it]

                   all         52         69      0.965      0.977      0.983       0.83



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      18/60      4.31G     0.6629     0.5014      1.014         25        640: 100%|██████████| 29/29 [00:27<00:00,  1.06it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:03<00:00,  1.50s/it]

                   all         52         69      0.962      0.929      0.961        0.8



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      19/60       4.3G     0.7025     0.5274      1.037         23        640: 100%|██████████| 29/29 [00:42<00:00,  1.45s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:03<00:00,  1.62s/it]

                   all         52         69      0.984      0.948      0.971       0.84



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      20/60       4.3G     0.6455     0.4689     0.9943         36        640: 100%|██████████| 29/29 [00:44<00:00,  1.54s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.44s/it]

                   all         52         69      0.946      0.957      0.965      0.826



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      21/60      4.28G     0.6669     0.4753      1.012         32        640: 100%|██████████| 29/29 [00:26<00:00,  1.09it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:03<00:00,  1.73s/it]

                   all         52         69      0.992      0.975      0.986      0.843



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      22/60       4.3G     0.6556     0.4866      1.013         24        640: 100%|██████████| 29/29 [00:28<00:00,  1.03it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.32s/it]

                   all         52         69      0.968      0.954      0.958      0.835



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      23/60      4.29G      0.654      0.465      1.011         36        640: 100%|██████████| 29/29 [00:45<00:00,  1.57s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.38s/it]

                   all         52         69      0.961      0.978      0.986      0.855



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      24/60      4.31G     0.6427      0.456      1.004         33        640: 100%|██████████| 29/29 [00:48<00:00,  1.66s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.47s/it]

                   all         52         69      0.974      0.975       0.98      0.855



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      25/60       4.3G     0.6069     0.4534     0.9937         31        640: 100%|██████████| 29/29 [00:29<00:00,  1.00s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:03<00:00,  1.56s/it]

                   all         52         69      0.977      0.966      0.984      0.847



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      26/60       4.3G     0.6365      0.464      1.011         30        640: 100%|██████████| 29/29 [00:44<00:00,  1.54s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.50s/it]

                   all         52         69      0.996      0.966      0.984      0.856



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      27/60       4.3G     0.6362     0.4614      1.011         31        640: 100%|██████████| 29/29 [00:47<00:00,  1.65s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.44s/it]

                   all         52         69      0.998      0.955      0.985      0.872



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      28/60       4.3G     0.6107     0.4238     0.9908         23        640: 100%|██████████| 29/29 [00:47<00:00,  1.64s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:03<00:00,  1.57s/it]

                   all         52         69      0.989      0.974      0.989      0.896



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      29/60       4.3G     0.5825     0.4151     0.9804         31        640: 100%|██████████| 29/29 [00:30<00:00,  1.06s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:03<00:00,  1.71s/it]

                   all         52         69      0.995      0.966      0.979      0.891



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      30/60      4.31G      0.592       0.42     0.9869         31        640: 100%|██████████| 29/29 [00:29<00:00,  1.03s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.36s/it]

                   all         52         69      0.997      0.977      0.988      0.872



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      31/60      4.29G     0.5748      0.409     0.9762         29        640: 100%|██████████| 29/29 [00:46<00:00,  1.59s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.30s/it]

                   all         52         69      0.995      0.977      0.989      0.888



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      32/60      4.31G      0.584     0.4177      0.982         30        640: 100%|██████████| 29/29 [00:36<00:00,  1.26s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.21s/it]

                   all         52         69      0.998      0.977      0.987      0.868



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      33/60       4.3G     0.5655     0.3982     0.9657         29        640: 100%|██████████| 29/29 [00:28<00:00,  1.01it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.20s/it]

                   all         52         69       0.99      0.976      0.987      0.879



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      34/60      4.31G     0.5651     0.3962     0.9673         32        640: 100%|██████████| 29/29 [00:26<00:00,  1.10it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.11s/it]

                   all         52         69      0.996      0.977      0.987      0.866



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      35/60       4.3G     0.5781     0.3907     0.9818         33        640: 100%|██████████| 29/29 [00:40<00:00,  1.40s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:03<00:00,  1.52s/it]

                   all         52         69      0.996      0.977      0.987      0.866



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      36/60      4.31G     0.5456     0.3814     0.9712         29        640: 100%|██████████| 29/29 [00:33<00:00,  1.15s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.27s/it]

                   all         52         69      0.998      0.977      0.988      0.888



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      37/60       4.3G     0.5342     0.3819     0.9617         26        640: 100%|██████████| 29/29 [00:30<00:00,  1.06s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.23s/it]

                   all         52         69      0.997      0.977      0.988      0.884



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      38/60       4.3G     0.5438     0.3892      0.963         25        640: 100%|██████████| 29/29 [00:32<00:00,  1.12s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:03<00:00,  1.63s/it]

                   all         52         69      0.995      0.986      0.988      0.886



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      39/60       4.3G     0.5571     0.3886     0.9726         33        640: 100%|██████████| 29/29 [00:44<00:00,  1.52s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:03<00:00,  1.54s/it]

                   all         52         69      0.998      0.977       0.99      0.913



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      40/60      4.31G     0.5307     0.3587     0.9601         33        640: 100%|██████████| 29/29 [00:51<00:00,  1.79s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.36s/it]

                   all         52         69      0.977      0.988      0.989       0.89



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      41/60       4.3G     0.5422     0.3718     0.9561         33        640: 100%|██████████| 29/29 [00:32<00:00,  1.14s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.23s/it]

                   all         52         69      0.989      0.989       0.99      0.917



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      42/60      4.32G     0.5088     0.3619     0.9529         31        640: 100%|██████████| 29/29 [00:43<00:00,  1.51s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:03<00:00,  1.72s/it]

                   all         52         69      0.998      0.977      0.989      0.909



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      43/60      4.31G     0.4899     0.3408     0.9385         36        640: 100%|██████████| 29/29 [00:31<00:00,  1.09s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.45s/it]

                   all         52         69      0.999      0.978      0.991       0.89



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      44/60      4.31G     0.4953     0.3416     0.9408         20        640: 100%|██████████| 29/29 [00:47<00:00,  1.65s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.41s/it]

                   all         52         69      0.996      0.989      0.991      0.887



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      45/60      4.28G     0.4928     0.3314     0.9366         26        640: 100%|██████████| 29/29 [00:41<00:00,  1.44s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.31s/it]

                   all         52         69      0.995      0.988       0.99      0.892



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      46/60      4.31G     0.4558     0.3268     0.9241         17        640: 100%|██████████| 29/29 [00:30<00:00,  1.06s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:03<00:00,  1.76s/it]

                   all         52         69      0.999      0.978      0.991      0.905



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      47/60      4.29G     0.4621     0.3151     0.9249         33        640: 100%|██████████| 29/29 [00:29<00:00,  1.02s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.29s/it]

                   all         52         69      0.998      0.978       0.99        0.9



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      48/60      4.31G     0.4605       0.31     0.9355         33        640: 100%|██████████| 29/29 [00:35<00:00,  1.22s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.22s/it]

                   all         52         69      0.998      0.989       0.99      0.908



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      49/60      4.28G     0.4785     0.3239     0.9453         27        640: 100%|██████████| 29/29 [00:39<00:00,  1.37s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.28s/it]

                   all         52         69       0.98      0.989       0.99      0.907



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      50/60       4.3G     0.4615     0.3122     0.9246         34        640: 100%|██████████| 29/29 [00:30<00:00,  1.04s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.16s/it]

                   all         52         69       0.98      0.978       0.99      0.897


Closing dataloader mosaic

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      51/60       4.3G     0.3503     0.2364     0.8471         14        640: 100%|██████████| 29/29 [00:42<00:00,  1.45s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.36s/it]

                   all         52         69      0.999      0.978      0.989      0.894



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      52/60      4.31G     0.3503     0.2273     0.8409         16        640: 100%|██████████| 29/29 [00:44<00:00,  1.52s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:03<00:00,  1.53s/it]

                   all         52         69      0.997      0.978       0.99      0.917



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      53/60       4.3G      0.334     0.2197     0.8214         13        640: 100%|██████████| 29/29 [00:41<00:00,  1.45s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.40s/it]

                   all         52         69      0.999      0.978       0.99      0.914



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      54/60      4.32G      0.331     0.2166     0.8284         16        640: 100%|██████████| 29/29 [00:32<00:00,  1.11s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:03<00:00,  1.71s/it]

                   all         52         69      0.992      0.988       0.99      0.914



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      55/60       4.3G     0.3289     0.2044     0.8208         15        640: 100%|██████████| 29/29 [00:30<00:00,  1.06s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.36s/it]

                   all         52         69      0.993      0.988       0.99      0.918



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      56/60      4.31G     0.3231     0.2049     0.8272         16        640: 100%|██████████| 29/29 [00:52<00:00,  1.80s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.32s/it]

                   all         52         69      0.996      0.982       0.99      0.912



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      57/60       4.3G     0.3201     0.1998     0.8183         15        640: 100%|██████████| 29/29 [00:41<00:00,  1.45s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.35s/it]

                   all         52         69      0.999      0.979      0.991      0.915



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      58/60      4.32G     0.3209     0.1954     0.8102         15        640: 100%|██████████| 29/29 [00:43<00:00,  1.51s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:03<00:00,  1.53s/it]

                   all         52         69      0.999      0.978      0.991      0.913



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      59/60       4.3G     0.3032     0.1962     0.8118         14        640: 100%|██████████| 29/29 [00:28<00:00,  1.02it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:03<00:00,  1.55s/it]

                   all         52         69      0.999      0.978      0.991      0.905



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      60/60      4.29G     0.3015     0.1889     0.8272         12        640: 100%|██████████| 29/29 [00:44<00:00,  1.52s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:02<00:00,  1.30s/it]

                   all         52         69      0.999      0.978      0.991      0.919



60 epochs completed in 0.714 hours.
Optimizer stripped from Z:\Proyectos\Robotica\footbot\soccer_vision\runs\yolo11s_train\weights\last.pt, 19.2MB
Optimizer stripped from Z:\Proyectos\Robotica\footbot\soccer_vision\runs\yolo11s_train\weights\best.pt, 19.2MB

Validating Z:\Proyectos\Robotica\footbot\soccer_vision\runs\yolo11s_train\weights\best.pt...
Ultralytics 8.3.0  Python-3.11.5 torch-2.5.1+cu121 CUDA:0 (NVIDIA GeForce RTX 3050 Ti Laptop GPU, 4096MiB)
YOLO11s summary (fused): 238 layers, 9,413,574 parameters, 0 gradients, 21.3 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 2/2 [00:00<00:00,  2.36it/s]


                   all         52         69      0.999      0.978      0.991      0.918
                  goal         25         25      0.998          1      0.995      0.892
              opponent         43         44          1      0.956      0.987      0.944
Speed: 0.5ms preprocess, 7.9ms inference, 0.0ms loss, 1.7ms postprocess per image
Results saved to Z:\Proyectos\Robotica\footbot\soccer_vision\runs\yolo11s_train
[INFO] [OK] Saved best weights to Z:\Proyectos\Robotica\footbot\soccer_vision\models\yolo11s\soccer_yolo.pt
[INFO] [OK] Copied training artifacts to Z:\Proyectos\Robotica\footbot\soccer_vision\models\yolo11s\train_artifacts
[INFO] [PREDICT] Running quick predictions on val images…

image 1/52 Z:\Proyectos\Robotica\footbot\soccer_vision\dataset\val\images\0cd54e92-photo_2025-10-12_22-02-19_aug_1.jpg: 640x640 1 goal, 1 opponent, 43.5ms
image 2/52 Z:\Proyectos\Robotica\footbot\soccer_vision\dataset\val\images\16444e52-photo_2025-10-12_22-01-24_aug_2.jpg: 640x640 1 goa

<!-- === Results summary (pretty, compact) === -->

<style>
.results-wrap { font-family: system-ui, -apple-system, Segoe UI, Roboto, "Helvetica Neue", Arial, "Noto Sans", "Liberation Sans", sans-serif; }
.results-head { margin: 0 0 .6rem 0; font-size: 1.35rem; font-weight: 700; }
.results-sub { margin: 0 0 1.2rem 0; color: #6b7280; font-size: .95rem; }
.grid { display: grid; grid-template-columns: repeat(3, 1fr); gap: 10px; }
.grid-2 { display: grid; grid-template-columns: 1.2fr 1fr; gap: 12px; margin-top: 10px; }
.card { background:#0b0b0b0d; border:1px solid #e5e7eb; border-radius:12px; padding:10px; }
.card h4 { margin:.2rem 0 .5rem 0; font-size:.95rem; }
.fig { display:flex; flex-direction:column; align-items:center; gap:.35rem; }
.fig img { width:100%; height:auto; border-radius:8px; border:1px solid #e5e7eb; }
.caption { color:#6b7280; font-size:.85rem; line-height:1.2; }
ul.tight { margin:.4rem 0 .6rem 1.1rem; padding:0; }
ul.tight li { margin:.2rem 0; }
@media (max-width: 980px) {
  .grid { grid-template-columns: 1fr; }
  .grid-2 { grid-template-columns: 1fr; }
}
</style>

<div class="results-wrap">

<div class="results-head">✅ Training Results — Quick Visual Summary</div>
<div class="results-sub">Figures generated by Ultralytics and saved to <code>soccer_vision/results/</code>.</div>

<div class="grid">

<div class="card fig">
  <h4>F1–Confidence Curve</h4>
  <img src="../results/F1_curve.png" alt="F1 curve">
  <div class="caption">
    • F1 balances precision & recall as the confidence threshold changes.<br>
    • Use the peak as a good default threshold (in our run ≈ <b>0.86</b>).
  </div>
</div>

<div class="card fig">
  <h4>Precision–Recall (PR) Curve</h4>
  <img src="../results/PR_curve.png" alt="PR curve">
  <div class="caption">
    • Area near top-right means strong performance.<br>
    • Legend shows per-class AP and overall mAP@0.5.
  </div>
</div>

<div class="card fig">
  <h4>Recall–Confidence Curve</h4>
  <img src="../results/R_curve.png" alt="R curve">
  <div class="caption">
    • Recall starts high and drops as threshold increases.<br>
    • Explains why F1 peaks before very high thresholds.
  </div>
</div>

</div>

<div class="grid-2">

<div class="card fig">
  <h4>Normalized Confusion Matrix</h4>
  <img src="../results/confusion_matrix_normalized.png" alt="Confusion matrix normalized">
  <div class="caption">
    • Rows = True class, Columns = Predicted class (each row sums to 1).<br>
    • Diagonal cells are correct predictions (higher is better).<br>
    • Off-diagonals show typical confusions (e.g., <i>opponent → background</i>).
  </div>
</div>

<div class="card">
  <h4>How to read <code>results.png</code></h4>
  <div class="fig"><img src="../results/results.png" alt="Training curves grid"></div>
  <ul class="caption tight">
    <li><b>train/box_loss, train/cls_loss, train/dfl_loss</b> — training losses trending down = learning.</li>
    <li><b>val/box_loss, val/cls_loss, val/dfl_loss</b> — validation losses; should generally follow train (no big divergence).</li>
    <li><b>metrics/precision(B), metrics/recall(B)</b> — quality on validation set improves as they rise toward 1.0.</li>
    <li><b>metrics/mAP50(B)</b> & <b>mAP50-95(B)</b> — summary accuracy; smooth upward trend is ideal.</li>
  </ul>
  <div class="caption">
    If val metrics drop while train improves, consider more augmentation, earlier stopping, or smaller LR.
  </div>
</div>

</div>

<div class="card" style="margin-top:12px;">
  <h4>Takeaways & Tips</h4>
  <ul class="tight">
    <li><b>Default threshold:</b> start around the F1 peak (≈ <b>0.86</b>) and tune per use-case.</li>
    <li><b>Where are files?</b> <code>soccer_vision/results/</code> (or <code>models/&lt;subdir&gt;/train_artifacts/</code>).</li>
    <li><b>Best weights:</b> <code>models/&lt;subdir&gt;/soccer_yolo.pt</code> for inference.</li>
  </ul>
</div>

</div>
